Documentation/design-docs/removing-embedded-statements.md

```diff
@@ -0,0 +1,180 @@
+the same order in which they were originally performed.
+
+##### 3.ii. Lowering
+
+The picture for lowering is much the same as the picture for decomposition. Like decomposition, all rewrites are
+performed in execution order, and most rewrites are simple expansions of a single node into multiple nodes. There is one
+notable exception, however: the lowering of add nodes for the x86 architecture examines the parent stack provided by
+tree walks in order to defer lowering until it may be possible to form an address mode. In this case, the use of the
+parent stack can be replaced with a helper that finds the node that uses the def produced by the add node.
+
+##### 3.iii. General issues
+
+Both decomposition and lowering make use of a utility to fix up information present in the per-call-node side table that
+tracks extra information about the call’s arguments when necessary. This helper can be replaced by instead fixing up the
+side table entries when the call is visited by each pass.
+
+#### 4. Remove uses of statements and the tree-order invariant from LSRA.
+
+LSRA currently depends on the tree order invariant so that it can build a stack to track data associated with the defs
+consumed by an operator. Removing the tree order invariant requires that LSRA is able to accommodate situations where
+the contents of this stack as produced by a linear walk would no longer contain correct information due to the insertion
+of new nodes that produce values that are live across the operator and some subset of its operands. Analysis indicates
+that a simple map from defs to the necessary information should be sufficient.
+
+#### 5. Remove uses of statements from the rest of the backend.
+
+The rest of the backend--i.e. codegen--has no known dependencies on tree order, so there are no concerns with respect to
+the change in ordering semantics. However, statement nodes are used to derive IL offsets for debug information. There
+are a number of alternative methods of tracking IL offsets to consider:
+
+1.  Insert “IL offset” nodes into the linearly-ordered LIR. These nodes would then be noted by code generation and used
+    to emit the necessary IP-mapping entries in the same way that statements are today. This has the disadvantage of
+    requiring additional work when performing code motion on LIR if the IL offset for a particular node needs to be
+    kept up-to-date. However, today’s backend does not perform this sort of code motion and even if it did, optimized
+    debugging is not currently a scenario, so a loss of debugging fidelity may be acceptable.
+
+2.  Use a side table to map from each node to its IL offset. Although this increases the total working set of the
+    backend, this has the advantage of making it easy to keep IL offsets correct in the face of code motion.
+
+3.  Add IL offset information directly to each node. This has the disadvantage of increasing the working set of the
+    entire compiler.
```

Comment:
The IL offset is a 4 byte integer. It happens so that `GenTree` has 4 bytes of alignment padding, if we re-arrange the data members a bit we might be able to use those bytes.

Current GenTree layout:
```
UserDefinedType: GenTree
Data           :   this+0x0, Member, Type: enum genTreeOps, gtOper
Data           :   this+0x1, Member, Type: enum var_types, gtType
Data           :   this+0x2, Member, Type: signed char, gtCSEnum
1 byte hole
Data           :   this+0x4, Member, Type: unsigned short, gtAssertionNum
Data           :   this+0x6, Member, Type: unsigned char, _gtCostEx
Data           :   this+0x7, Member, Type: unsigned char, _gtCostSz
Data           :   this+0x8, Member, Type: unsigned char, _gtRegNum
3 bytes hole
Data           :   this+0xC, Member, Type: unsigned int, gtFlags
Data           :   this+0x10, Member, Type: struct ValueNumPair, gtVNPair
Data           :   this+0x18, Member, Type: unsigned int, gtRsvdRegs
Data           :   this+0x1C, Member, Type: class TreeNodeInfo, gtLsraInfo
Data           :   this+0x28, Member, Type: struct GenTree *, gtNext
Data           :   this+0x30, Member, Type: struct GenTree *, gtPrev
```